### PR TITLE
Feat/Return to the most prioritized morph endpoint

### LINKED ISSUE #1615: Consider automatic swtich to the morph RPC with the highest priority
After our main endpoint has gone down, we would like to eventually switch back to it even if there are no errors with the other endpoint.

### PR DESCRIPTION
**Pull request by carpawell**
Closes #1615.

The switching period could be configurable OR could be kinda flexible (try to wait more time if a node refuses connection more than one time) but it complicates the code and I do not think that it is worth it.
Also, we could not just try to connect to the most prioritized RPC but simply increase priority if the most prioritized RPC does not allow a connection. Not sure if we need it.

Checked using https://github.com/carpawell/neofs-dev-env/tree/local/add-rpc-morph-node.

### CODE REVIEW

**Comment by carpawell** [comment]
Too dangerous PR, waits `v0.32.0` release.

File: pkg/morph/client/multi.go [hunk 1]
```diff
@@ -120,6 +132,79 @@ func (c *Client) notificationLoop() {
 	}
 }
 
+func (c *Client) switchToMostPrioritized() {
+	const period = 30 * time.Second
```
Comment (fyrchik) [inline pkg/morph/client/multi.go:136]: I think 30 second is a really small time for this kind of thing, something like 10 minutes is more suitable IMO. cc @realloc

File: pkg/morph/client/multi.go [hunk 2]
```diff
@@ -120,6 +132,79 @@ func (c *Client) notificationLoop() {
...
+			for i, e := range c.endpoints.list {
+				if i == c.endpoints.curr {
+					continue
+				}
+
+				if e.Priority < currPriority {
+					tryE = e.Address
+					break
+				}
+			}
```
Comment (fyrchik) [inline pkg/morph/client/multi.go:159]: Endpoints are already sorted by priority, so we can just try each node in list until the current one.
This also helps in a situation when we failed to connect to the most prioritized node: in this case the next should be tried in the same loop.

File: pkg/morph/client/multi.go [hunk 3]
```diff
@@ -63,6 +61,16 @@ func (c *Client) switchRPC() bool {
...
 		}
 
+		c.client = cli
+		c.rpcActor = act
+		c.gasToken = gas
+
+		if !c.switchIsActive.Load() &&
+			c.endpoints.list[c.endpoints.curr].Priority != c.endpoints.list[0].Priority {
+			c.switchIsActive.Store(true)
+			go c.switchToMostPrioritized()
```
Comment (fyrchik) [inline pkg/morph/client/multi.go:71]: Do we create a goroutine each time `switchRPC` is called? I think this should be a background worker.

File: pkg/morph/client/multi.go (hunk 1 above)
Comment (realloc) [reply pkg/morph/client/multi.go:136]: I'd suggest moving all similar constants to a config file anyway.

Comment (fyrchik) [reply pkg/morph/client/multi.go:136]: Agree, but the default still should make sense.

Comment (realloc) [reply pkg/morph/client/multi.go:136]: From the test cases, we often see 2 minutes as a requirement.

File: pkg/morph/client/multi.go (hunk 3 above)
Comment (carpawell) [reply pkg/morph/client/multi.go:71]: > Do we create a goroutine each time switchRPC is called?

yes if it has not been created yet: my idea was that a node with a non-prioritized connection is smth really rare, so i do not want to have a goroutine if i know that i do not need it

sounds strange but that was a 1-month-younger me. what do you think?

File: pkg/morph/client/multi.go (hunk 1 above)
Comment (carpawell) [reply pkg/morph/client/multi.go]: changed to 2 min

> moving all similar constants to a config file anyway

should it be done in that PR? not sure if it is useful. mb add when it has use cases?

File: pkg/morph/client/multi.go (hunk 2 above)
Comment (carpawell) [reply pkg/morph/client/multi.go]: > Endpoints are already sorted by priority, so we can just try each node in list until the current one.

If we have `[a, b, c, d]` nodes with `[1, 2, 2, 2]` priorities and are connected to `d`, it is useless for us to reconnect to any node that is not `a`. At least this is how i understand the priorities

> we failed to connect to the most prioritized node: in this case the next should be tried in the same loop

well, that is one of the questions i was not sure about. ok, added that

File: pkg/morph/client/multi.go (hunk 1 above)
Comment (fyrchik) [reply pkg/morph/client/multi.go]: It is definitely useful, because we must be able to control such things.

Comment (carpawell) [reply pkg/morph/client/multi.go]: added

File: pkg/morph/client/multi.go [hunk 4]
```diff
@@ -120,6 +132,88 @@ func (c *Client) notificationLoop() {
...
+			copy(endpointsCopy, c.endpoints.list)
+
+			currEndpoint := c.endpoints.curr
+			currPriority := c.endpoints.list[c.endpoints.curr].Priority
+			highestPriority := c.endpoints.list[0].Priority
+			c.switchLock.RUnlock()
+
+			if currPriority == highestPriority {
+				// already connected to
+				// the most prioritized
```
Comment (fyrchik) [inline pkg/morph/client/multi.go:156]: What about doing it in `defer`?

**Review by acid-ant — APPROVED** [review]

File: docs/storage-node-configuration.md [hunk 5]
```diff
@@ -132,18 +132,20 @@ contracts:
...
+|-------------------|-----------------------------------------------------------|------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `dial_timeout`    | `duration`                                                | `5s`             | Timeout for dialing connections to N3 RPCs.                                                                                                                         |
+| `cache_ttl`       | `duration`                                                | Morph block time | Sidechain cache TTL value (min interval between similar calls).<br/>Negative value disables caching.<br/>Cached entities: containers, container lists, eACL tables. |
+| `switch_interval` | `duration`                                                | `2m`             | Time interval between the attempts to connect to the highest priority RPC node if the connection is not established yet.                                            |
```
Comment (fyrchik) [inline docs/storage-node-configuration.md:147]: `switch` is alphabetially greater than `rpc_endpoint`, make it the last?

File: pkg/morph/client/constructor.go [hunk 6]
```diff
@@ -60,6 +63,7 @@ func defaultConfig() *cfg {
 		signer: &transaction.Signer{
 			Scopes: transaction.Global,
 		},
+		switchInterval: defaultSwitchInterval,
```
Comment (fyrchik) [inline pkg/morph/client/constructor.go:66]: Can we make it `0` (disabled) in `morph/client` and ensure to provide interval in both IR and neofs-node?

File: pkg/morph/client/multi.go [hunk 7]
```diff
@@ -120,6 +132,88 @@ func (c *Client) notificationLoop() {
...
+				c.switchIsActive.Store(false)
+				return
+			}
+
+			for i, e := range endpointsCopy {
+				if i == currEndpoint || currPriority == e.Priority {
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/morph/client/multi.go:162]: part of the old code, dropped

File: pkg/morph/client/multi.go [hunk 8]
```diff
@@ -120,6 +132,88 @@ func (c *Client) notificationLoop() {
...
+
+			currEndpoint := c.endpoints.curr
+			currPriority := c.endpoints.list[c.endpoints.curr].Priority
+			highestPriority := c.endpoints.list[0].Priority
+			c.switchLock.RUnlock()
+
+			if currPriority == highestPriority {
+				// already connected to
+				// the most prioritized
+				c.switchIsActive.Store(false)
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/morph/client/multi.go:157]: good, added

**Pull request merged**